runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -114,7 +144,525 @@ native2InterpJavaUpcallD(J9UpcallMetaData *data, void *argsListPointer)
 U_8 * JNICALL
 native2InterpJavaUpcallStruct(J9UpcallMetaData *data, void *argsListPointer)
 {
-	return NULL;
+	UDATA returnValue = native2InterpJavaUpcallImpl(data, argsListPointer);
+	return (U_8 *)returnValue;
+}
+
+/**
+ * @brief Determine the predefined return type against the return signature type
+ * stored in the native signature array of the upcall metadata.
+ *
+ * @param data a pointer to J9UpcallMetaData
+ * @return a U_8 value for the return type
+ */
+static U_8
+getReturnTypeFromMetaData(J9UpcallMetaData *data)
+{
+	J9JavaVM *vm = data->vm;
+	J9VMThread *currentThread = currentVMThread(vm);
+	j9object_t methodType = J9VMOPENJ9INTERNALFOREIGNABIUPCALLMHMETADATA_CALLEETYPE(currentThread,
+			J9_JNI_UNWRAP_REFERENCE(data->mhMetaData));
+	J9Class *retClass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread,
+			J9VMJAVALANGINVOKEMETHODTYPE_RTYPE(currentThread, methodType));
+	J9UpcallNativeSignature *nativeSig = data->nativeFuncSignature;
+	J9UpcallSigType *sigArray = nativeSig->sigArray;
+	/* The last element is for the return type */
+	U_8 retSigType = sigArray[nativeSig->numSigs - 1].type & J9_FFI_UPCALL_SIG_TYPE_MASK;
+	U_8 returnType = 0;
+
+	switch (retSigType) {
+	case J9_FFI_UPCALL_SIG_TYPE_VOID:
+		returnType = J9NtcVoid;
+		break;
+	case J9_FFI_UPCALL_SIG_TYPE_CHAR:
+		returnType = (retClass == vm->booleanReflectClass) ? J9NtcBoolean : J9NtcByte;
+		break;
+	case J9_FFI_UPCALL_SIG_TYPE_SHORT:
+		returnType = (retClass == vm->charReflectClass) ? J9NtcChar : J9NtcShort;
+		break;
+	case J9_FFI_UPCALL_SIG_TYPE_INT32:
+		returnType = J9NtcInt;
+		break;
+	case J9_FFI_UPCALL_SIG_TYPE_INT64:
+		returnType = J9NtcLong;
+		break;
+	case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+		returnType = J9NtcFloat;
+		break;
+	case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+		returnType = J9NtcDouble;
+		break;
+	case J9_FFI_UPCALL_SIG_TYPE_POINTER:
+		returnType = J9NtcPointer;
+		break;
+	case J9_FFI_UPCALL_SIG_TYPE_STRUCT:
+		returnType = J9NtcStruct;
+		break;
+	default:
+		Assert_VM_unreachable();
+		break;
+	}
+
+	return returnType;
+}
+
+/**
+ * @brief The common helper function that calls into the interpreter
+ * via native2InterpJavaUpcallImpl to invoke the OpenJDK MH in the upcall
+ *
+ * @param data a pointer to J9UpcallMetaData
+ * @param argsListPointer a pointer to the argument list
+ * @return the expected value against the specified return type
+ */
+static UDATA JNICALL
+native2InterpJavaUpcallImpl(J9UpcallMetaData *data, void *argsListPointer)
+{
+	J9JavaVM *vm = data->vm;
```

Comment:
I see the following build error with this variable.  Please remove this line.

```
/Users/openj9/openj9-openjdk-jdk19/openj9/runtime/vm/UpcallVMHelpers.cpp:220:12: error: unused variable 'vm' [-Werror,-Wunused-variable]
        J9JavaVM *vm = data->vm;
                  ^
```

Reply:
Fixed.